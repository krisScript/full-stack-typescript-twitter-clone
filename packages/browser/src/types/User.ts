import CommonUser from '@twtr/common/source/types/user';

export default interface User extends CommonUser {
  following: string[];
  likes: string[];
  bookmarks: string[];
  replies: string[];
  retweets: string[];
  avatar?: string;
  _id: string;
}
