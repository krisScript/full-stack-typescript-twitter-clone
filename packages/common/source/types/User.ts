export default interface User {
  username: string;
  handle: string;
  email: string;
  website: string | undefined;
  profilePhoto: string;
  headerPhoto: string;
  date: string;
  followers: number;
}
