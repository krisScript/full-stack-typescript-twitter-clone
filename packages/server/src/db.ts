import mongoose from "mongoose";

const mongoURI: string = `mongodb+srv://${process.env.MONGO_USER}:${process.env.MONGO_PASSWORD}@cluster0-zmcyw.mongodb.net/${process.env.MONGO_DATABASE}?retryWrites=true`;
const db = async (): Promise<void> => {
  try {
    await mongoose.connect(mongoURI, { useNewUrlParser: true });
  } catch (err) {}
};

export default db;
