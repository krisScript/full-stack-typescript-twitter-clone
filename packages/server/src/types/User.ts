import CommonUser from '@twtr/common/source/types/User';
import { Schema, Document } from 'mongoose';

export default interface User extends CommonUser, Document {
  password: string;
  confirmed: boolean;
  following: Schema.Types.ObjectId[];
  likes: Schema.Types.ObjectId[];
  bookmarks: Schema.Types.ObjectId[];
}
