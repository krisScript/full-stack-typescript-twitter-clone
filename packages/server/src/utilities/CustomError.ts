import ValidationError from '@twtr/common/types/ValidationError';

const errors = {
  BadRequest: {
    status: 400,
    message: 'Request has wrong format.',
  },
  Unauthorized: {
    status: 401,
    message: 'Authentication credentials not valid.',
  },
  Forbidden: {
    status: 403,
    message: "You're missing permission to execute this request.",
  },
  NotFound: {
    status: 404,
    message: 'Resource not found.',
  },
  UnprocessableEntity: {
    status: 422,
    message: 'Request cannot be processed.',
  },
  InternalServerError: {
    status: 500,
    message: 'Internal server error.',
  },
};

class CustomError extends Error {
  public status: number;

  public message: string;

  public data?: ValidationError[] | string;

  public constructor(
    status: number,
    message: string,
    data?: ValidationError[] | string,
  ) {
    super();
    Object.setPrototypeOf(this, CustomError.prototype);
    this.status = status;
    this.message = message;
    this.data = data;
  }
}

export { CustomError, errors };
