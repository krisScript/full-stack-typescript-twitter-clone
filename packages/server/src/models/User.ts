import mongoose, { Schema } from 'mongoose';
import User from '@customTypes/User';

const UserSchema: Schema = new Schema({
  username: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  handle: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  email: { required: true, type: String, minlength: 3, unique: true },
  password: { required: true, type: String, minlength: 12 },
  website: { required: false, type: String, minlength: 3 },
  confirmed: { type: Boolean, default: false },
  profilePhoto: { type: String },
  headerPhoto: { type: String },
  date: {
    type: Date,
    default: Date.now,
  },
  following: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  followers: {
    type: Number,
    default: 0,
  },
  likes: [
    {
      source: {
        type: mongoose.Types.ObjectId,
        required: true,
        refPath: 'likes.ref',
      },
      ref: {
        type: String,
        required: true,
        enum: ['Tweet', 'Reply'],
      },
    },
  ],
  bookmarks: [
    {
      source: {
        type: mongoose.Types.ObjectId,
        required: true,
        refPath: 'bookmarks.ref',
      },
      ref: {
        type: String,
        required: true,
        enum: ['Tweet', 'Reply'],
      },
    },
  ],
});

export default mongoose.model<User>('User', UserSchema);
