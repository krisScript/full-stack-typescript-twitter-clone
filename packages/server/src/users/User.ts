import mongoose, { Schema } from 'mongoose';
import User from '../types/User';
import duplicationErrorHandler from '../middleware/duplicationErrorHandler';
//@ts-ignore
import uniqueValidator from 'mongoose-unique-validator';

const UserSchema: Schema = new Schema({
  username: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  handle: {
    required: true,
    type: String,
    minlength: 1,
    maxlength: 50,
    unique: true,
  },
  email: { required: true, type: String, minlength: 3, unique: true },
  password: { required: true, type: String, minlength: 12 },
  isConfirmed: { type: Boolean, default: true },
  avatar: {
    type: String,
  },
  cover: { type: String },
  date: {
    type: Date,
    default: Date.now,
  },
  following: [
    {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },
  ],
  followers: {
    type: Number,
    default: 0,
  },
  likes: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
  retweets: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
  replies: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],

  bookmarks: [
    {
      type: Schema.Types.ObjectId,
      ref: 'Tweet',
    },
  ],
});

UserSchema.index({ handle: 'text' });

UserSchema.plugin(uniqueValidator);

export default mongoose.model<User>('User', UserSchema);
