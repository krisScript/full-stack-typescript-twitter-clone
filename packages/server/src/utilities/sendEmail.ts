import nodemailer from 'nodemailer';
import MailOptions from '@customTypes/MailOptions';
import { CustomError, errors } from '@utilities/CustomError';

const sendEmail = (mailOptions: MailOptions): void => {
  try {
    const { EMAIL, EMAIL_PASSWORD } = process.env;
    const transporter = nodemailer.createTransport({
      service: 'gmail',
      auth: {
        user: EMAIL,
        pass: EMAIL_PASSWORD,
      },
    });
    transporter.sendMail(mailOptions);
  } catch (err) {
    const { status, message } = errors.UnprocessableEntity;
    const error = new CustomError(status, message);
    throw error;
  }
};
export default sendEmail;
