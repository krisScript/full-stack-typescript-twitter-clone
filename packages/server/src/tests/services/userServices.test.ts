import mongoose from 'mongoose';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import {
  createUser,
  sendConfirmationEmail,
  sendPasswordResetEmail,
  checkCredentialsAvailability,
  comparePasswords,
  checkUserConfirmation,
  getUserById,
  getUserByEmail,
} from '@services/userServices';
import User from '@models/User';
import db from 'src/db';
import sendEmail from '@utilities/sendEmail';
import { CustomError, errors } from '@utilities/CustomError';

jest.mock('@utilities/sendEmail');
jest.mock('jsonwebtoken');

const mockToken = 'mockToken';
(jwt.sign as jest.Mock).mockReturnValue(mockToken);

describe('userServices', (): void => {
  beforeAll(
    async (): Promise<void> => {
      db();
      await User.deleteMany({}).exec();
    },
  );
  afterEach(
    async (): Promise<void> => {
      await User.deleteMany({}).exec();
    },
  );
  afterAll(
    async (): Promise<void> => {
      await mongoose.disconnect();
    },
  );
  const secret = process.env.SECRET;
  const username = 'username';
  const handle = 'testUserHandle';
  const email = 'testmail@mail.com';
  const password = 'testPassword';
  describe('createUser', (): void => {
    it(`should create a new user`, async (): Promise<void> => {
      expect.assertions(7);
      const hashMock = jest.spyOn(bcrypt, 'hash');
      const { userId } = await createUser(username, handle, email, password);
      expect(userId).toBeTruthy();
      const user = await User.findById(userId);
      if (!user) {
        return;
      }
      expect(user.username).toMatch(username);
      expect(user.handle).toMatch(handle);
      expect(user.email).toMatch(email);
      expect(user.password).not.toMatch(password);
      expect(hashMock).toHaveBeenCalledTimes(1);
      expect(hashMock).toHaveBeenCalledWith(password, 12);
      hashMock.mockRestore();
    });
  });
  it('should throw an error', async (): Promise<void> => {
    expect.assertions(1);
    await User.insertMany({ username, handle, email, password });
    await expect(
      createUser(username, handle, email, password),
    ).rejects.toThrow();
  });
  describe('getUserByEmail', (): void => {
    it(`should get a user`, async (): Promise<void> => {
      const newUser = new User({
        username,
        handle,
        email,
        password,
      });
      await newUser.save();
      const { user } = await getUserByEmail(email);
      if (!user) {
        return;
      }
      expect(user.username).toMatch(username);
      expect(user.handle).toMatch(handle);
      expect(user.email).toMatch(email);
    });

    it(`should throw an error`, async (): Promise<void> => {
      const { status, message } = errors.NotFound;
      const error = new CustomError(status, message);
      await expect(getUserByEmail(email)).rejects.toThrow(error);
    });
  });
  describe('getUserById', (): void => {
    it(`should get a user`, async (): Promise<void> => {
      expect.assertions(3);
      const newUser = new User({
        username,
        handle,
        email,
        password,
      });
      await newUser.save();
      const userId = newUser._id;
      const { user } = await getUserById(userId);
      if (!user) {
        return;
      }
      expect(user.username).toMatch(username);
      expect(user.handle).toMatch(handle);
      expect(user.email).toMatch(email);
    });

    it(`should throw an error`, async (): Promise<void> => {
      const { status, message } = errors.NotFound;
      const error = new CustomError(status, message);
      const userId = mongoose.Types.ObjectId().toString();
      await expect(getUserById(userId)).rejects.toThrow(error);
    });
  });
  describe('sendConfirmationEmail', (): void => {
    it(`should call sign and sendEmail`, async (): Promise<void> => {
      expect.assertions(4);
      const userId = mongoose.Types.ObjectId().toString();
      sendConfirmationEmail(userId, email);
      expect(jwt.sign).toHaveBeenCalledTimes(1);
      expect(jwt.sign).toHaveBeenCalledWith(
        {
          userId,
        },
        secret,
        { expiresIn: '1h' },
      );
      expect(sendEmail).toHaveBeenCalledTimes(1);
      expect(sendEmail).toMatchSnapshot();
    });
  });
  describe('sendPasswordRequestEmail', (): void => {
    it(`should call sign and sendEmail`, async (): Promise<void> => {
      expect.assertions(4);
      const userId = mongoose.Types.ObjectId().toString();
      sendPasswordResetEmail(userId, email);
      expect(jwt.sign).toHaveBeenCalledTimes(1);
      expect(jwt.sign).toHaveBeenCalledWith(
        {
          userId,
        },
        secret,
        { expiresIn: '1h' },
      );
      expect(sendEmail).toHaveBeenCalledTimes(1);
      expect(sendEmail).toMatchSnapshot();
    });
  });
  describe('checkCredentialsAvailability', (): void => {
    it('should throw a validationError error', async (): Promise<void> => {
      expect.assertions(1);

      await User.insertMany({ username, handle, email, password });
      await expect(
        checkCredentialsAvailability(username, handle, email),
      ).rejects.toThrow();
    });
    it('should not throw an error', async (): Promise<void> => {
      expect.assertions(1);
      await expect(
        checkCredentialsAvailability(username, handle, email),
      ).resolves.toBeUndefined();
    });
  });

  describe('comparePasswords', (): void => {
    let hashedPassword: string;
    beforeEach(
      async (): Promise<void> => {
        hashedPassword = await bcrypt.hash(password, 12);
      },
    );
    it(`should throw an error`, async (): Promise<void> => {
      const incorrectPassword = 'incorrectPassword';
      const { status, message } = errors.Unauthorized;
      const error = new CustomError(status, message);
      await expect(
        comparePasswords(incorrectPassword, hashedPassword),
      ).rejects.toThrow(error);
    });
    it(`should not throw an error`, async (): Promise<void> => {
      await expect(comparePasswords(password, hashedPassword)).toEqual(
        Promise.resolve({}),
      );
    });
  });
  describe('checkUserConfirmation', (): void => {
    it(`should throw an error`, async (): Promise<void> => {
      const user = new User({
        username,
        handle,
        email,
        password,
      });
      await user.save();
      const { status, message } = errors.Unauthorized;
      const error = new CustomError(status, message);
      await expect(checkUserConfirmation(user)).rejects.toThrow(error);
    });
    it(`should not throw an error`, async (): Promise<void> => {
      const user = new User({
        username,
        handle,
        email,
        password,
      });
      user.confirmed = true;
      await user.save();
      await expect(checkUserConfirmation(user)).toEqual(Promise.resolve());
    });
  });
});
