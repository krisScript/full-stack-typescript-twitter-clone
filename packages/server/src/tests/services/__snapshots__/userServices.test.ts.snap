// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userServices areCredentialsAvailable should throw an error when the user credentials are already taken 1`] = `[Error: Conflict]`;

exports[`userServices checkUserConfirmation shouldn't throw an error when the user confirmed his email address 1`] = `[Error: Authentication credentials not valid]`;

exports[`userServices checkUserConfirmation should throw an error the user hasn't confirmed his email address 1`] = `[Error: Authentication credentials not valid]`;

exports[`userServices getUserByEmail should throw an error when the user is not found 1`] = `[Error: Resource not found]`;

exports[`userServices getUserById should throw an error should throw an error when the user is not found 1`] = `[Error: Resource not found]`;
